src/lib/Microsoft.Health.Events/EventConsumers/Service/EventConsumerService.cs

```diff
@@ -35,44 +51,42 @@ public async Task ConsumeEvents(IEnumerable<IEventMessage> events)
             {
                 foreach (IEventConsumer eventConsumer in _eventConsumers)
                 {
-                    await OperationWithRetryAsync(eventConsumer, events);
+                    try
+                    {
+                        await RetryPolicy.ExecuteAsync(async () => await eventConsumer.ConsumeAsync(events).ConfigureAwait(false));
+                    }
+                    catch (Exception e)
+                    {
+                        _logger.LogError(e);
+                    }
                 }
             }
         }
 
-        private async Task OperationWithRetryAsync(IEventConsumer eventConsumer, IEnumerable<IEventMessage> events)
+        private static AsyncPolicy CreateRetryPolicy(ITelemetryLogger logger)
         {
-            int currentRetry = 0;
-            double backoffMs = 0;
-            Random random = new Random();
-            bool operationComplete = false;
-
-            while (!operationComplete)
+            bool ExceptionRetryableFilter(Exception ee)
             {
-                try
+                switch (ee)
                 {
-                    if (currentRetry > 0 && backoffMs < _maximumBackoffMs)
-                    {
-                        int randomMs = random.Next(0, 1000);
-                        backoffMs = Math.Pow(2000, currentRetry) + randomMs;
-                        await Task.Delay((int)backoffMs);
-                    }
-
-                    await TryOperationAsync(eventConsumer, events).ConfigureAwait(false);
-                    break;
-                }
-#pragma warning disable CA1031
-                catch (Exception e)
-#pragma warning restore CA1031
-                {
-                    _logger.LogError(e);
+                    case AggregateException ae when ae.InnerExceptions.All(ExceptionRetryableFilter):
```

Comment:
Should this be Any()? Basically in a mixed exception setting (some retryable some fatal) this will be cause the retryable exceptions to be skipped.  Shouldn't we retry the whole batch unless only fatal errors remain then bubble up?

Reply:
Yea that makes sense. I'll change this.